### PR TITLE
8310220: IGV: dump graph after each IGVN step at level 4

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset instruments Iterative GVN (IGVN) in C2 to dump the Ideal graph after each effective step (i.e. when the graph is rewritten or the recorded types are refined). This enables fine-grained tracing of IGVN transformation sequences using Ideal Graph Visualizer. This technique has proved useful for the investigation of [JDK-8303513](https://bugs.openjdk.org/browse/JDK-8303513), and can be also useful for educational purposes:

igv-level4

These new dumps are emitted at print level 4 (`PrintIdealGraphLevel=4`), the highest level of detail.

Following [feedback](https://bugs.openjdk.org/browse/JDK-8310220?focusedCommentId=14590132&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14590132) and offline discussions with Christian Hagedorn, the changeset also dumps the Ideal graph before and after IGVN at print level 3. This makes it possible to identify the source of graph changes between IGVN and other phases such as loop transformations. The existing phase `PHASE_MACH_ANALYSIS` is also promoted to print level 3, since it prints a single graph per compilation unit only (see print level documentation updates in this changeset). These additional changes increase the number of graph dumps per compilation at print level 3 by around 1.5x:

igv-level3

Finally, the verbose and rarely used bytecode parsing dumps are relegated to a new print level 5, which leaves the number of graphs per compilation at level 4 roughly as before the changeset.

#### Testing

- tier1-3 (linux-x64; release and debug mode).

- Verified that thousands of new IGVN graph dumps are correctly opened and visualized with the Ideal Graph Visualizer, at print levels 3 to 5.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310220](https://bugs.openjdk.org/browse/JDK-8310220): IGV: dump graph after each IGVN step at level 4 (**Enhancement** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Holenstein](https://openjdk.org/census#tholenstein) (@tobiasholenstein - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14537/head:pull/14537` \
`$ git checkout pull/14537`

Update a local copy of the PR: \
`$ git checkout pull/14537` \
`$ git pull https://git.openjdk.org/jdk.git pull/14537/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14537`

View PR using the GUI difftool: \
`$ git pr show -t 14537`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14537.diff">https://git.openjdk.org/jdk/pull/14537.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14537#issuecomment-1597039494)

### CODE REVIEW

**Comment by robcasloz** [comment]
Putting this changeset on hold, since I will not be able to react to reviewers' feedback in the next weeks.

File: src/hotspot/share/opto/phaseX.cpp [hunk 1]
```diff
@@ -891,9 +891,12 @@ void PhaseIterGVN::verify_step(Node* n) {
 }
 
 void PhaseIterGVN::trace_PhaseIterGVN(Node* n, Node* nn, const Type* oldtype) {
+  const Type* newtype = type_or_null(n);
+  if (nn != n || oldtype != newtype) {
+    C->print_method(PHASE_AFTER_ITER_GVN_STEP, 4, n);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phaseX.cpp:896]: Should we keep this at level 4 and move the parser generated dumps to a new level 5? I often add dumps for these steps during IGVN but I rarely ever need the parser generated dumps.

Comment (robcasloz) [reply src/hotspot/share/opto/phaseX.cpp:896]: Thanks for the feedback Christian, your suggestion makes sense. I will address this in a few weeks when I am back from vacation.

Comment (chhagedorn) [reply src/hotspot/share/opto/phaseX.cpp:896]: Sounds good, thanks Roberto and enjoy your vacation!

Comment (robcasloz) [reply src/hotspot/share/opto/phaseX.cpp:896]: Done! Please re-review.

**Comment by robcasloz** [comment]
> That looks good to me! Thanks for introducing the new level for the parsing dumps.

Thanks for reviewing, Christian!

**Review by tobiasholenstein — APPROVED** [review]
Looks good to me.

Thanks for adding this @robcasloz !

**Comment by robcasloz** [comment]
> Looks good to me.
> 
> Thanks for adding this @robcasloz !

Thanks for reviewing, Toby!

**Pull request closed without merging**